**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №13

Набор данных: movies.csv

Атрибуты:
title - Название фильма
release_year - Год выпуска
score - Рейтинг
number_of_votes - Количество оценок
duration - Продолжительность
main_genre - Основной жанр
main_production - Страна производства

### 1.Чтение файла (набора данных)

In [4]:
import pandas
db = pandas.read_csv('movies.csv', sep=';')

In [ ]:
# импорт библиотек, чтение файла с помощью pandas

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [6]:
db.head(20)

title  release_year  score  \
0      David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                     Inception        2010.0    8.8   
2                                  Forrest Gump        1994.0    8.8   
3                                    Anbe Sivam        2003.0    8.7   
4                            Bo Burnham: Inside        2021.0    8.7   
5                           Saving Private Ryan        1998.0    8.6   
6                              Django Unchained        2012.0    8.4   
7                                        Dangal        2016.0    8.4   
8                        Bo Burnham: Make Happy        2016.0    8.4   
9                         Louis C.K.: Hilarious        2010.0    8.4   
10              Dave Chappelle: Sticks & Stones        2019.0    8.4   
11                                     3 Idiots        2009.0    8.4   
12                                 Black Friday        2004.0    8.4   
13                                 Super Deluxe        2019.0    8.4   
14  Winter on Fire: Ukraine's Fight for Freedom        2015.0    8.3   
15                  Once Upon a Time in America        1984.0    8.3   
16                                  Taxi Driver        1976.0    8.3   
17                          Like Stars on Earth        2007.0    8.3   
18                            Bo Burnham: What.        2013.0    8.3   
19                            Full Metal Jacket        1987.0    8.3   

    number_of_votes  duration   main_genre main_production  
0           31180.0        83  documentary              GB  
1         2268288.0       148        scifi              GB  
2         1994599.0       142        drama              US  
3           20595.0       160       comedy              IN  
4           44074.0        87       comedy              US  
5         1346020.0       169        drama              US  
6         1472668.0       165      western              US  
7          180247.0       161       action              IN  
8           14356.0        60       comedy              US  
9           11973.0        84       comedy              US  
10          25687.0        65       comedy              US  
11         385782.0       170       comedy              IN  
12          20611.0       143        crime              IN  
13          13680.0       176     thriller              IN  
14          17710.0        98  documentary              UA  
15         342335.0       229        drama              US  
16         795222.0       113        crime              US  
17         188234.0       165        drama              IN  
18          11488.0        60       comedy              US  
19         723306.0       116        drama              GB

In [5]:
# применен метод head

2.2 Оценка данных с помощью метода info.

In [7]:
db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            384 non-null    object 
 1   release_year     387 non-null    float64
 2   score            386 non-null    float64
 3   number_of_votes  388 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  388 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.4+ KB


In [ ]:
# выполнен метод info

2.3 Оценка данных с помощью метода describe.

In [8]:
db.describe()

release_year       score  number_of_votes    duration
count    387.000000  386.000000     3.880000e+02  389.000000
mean    2010.976744    7.512176     1.391523e+05  123.352185
std       10.805726    0.443870     2.365279e+05   28.304800
min     1954.000000    6.900000     1.013900e+04   28.000000
25%     2007.500000    7.100000     2.055375e+04  104.000000
50%     2014.000000    7.400000     4.606950e+04  122.000000
75%     2018.000000    7.800000     1.551585e+05  139.000000
max     2022.000000    9.000000     2.268288e+06  229.000000

In [ ]:
# оценены числовые столбцы с помощью describe


---

**Датасет содержит информацию о 389 фильмах с 2010-х годов (медианный год выпуска - 2014), где большинство картин имеют высокие рейтинги (среднее 7.5/10) при стандартной продолжительности около 2 часов. Количество оценок варьируется от 10 тысяч до 2.3 миллионов, что указывает на смесь нишевых и популярных фильмов, при этом основные жанры представлены драмами и комедиями из разных стран, преимущественно США и Индии.**


 ---



 2.4 Оценка названий столбцов

In [10]:
db.columns

Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')

In [ ]:
# Выведены на экран названия столбцов с помощью df.columns. Проблемы не выявлены - названия столбцов соответствуют содержащейся в них информаци

### 3. Проверка пропусков

In [12]:
db_nan = db[db.isnull().any(axis=1)]
display(db_nan)

title  release_year  score  number_of_votes  duration main_genre  \
121    NaN        2017.0    7.8          25593.0        94     comedy   
141    NaN        2015.0    7.4          13703.0       101    musical   
174    NaN           NaN    NaN              NaN       121      drama   
187    NaN        2013.0    7.5          93314.0       128      drama   
214  Crazy           NaN    NaN         507878.0       118    romance   
226    NaN        1999.0    NaN         180532.0       127      drama   

    main_production  
121              US  
141              US  
174             NaN  
187              US  
214              US  
226              US

In [ ]:
# Проверены данные на наличие пропусков. 

In [13]:
db = db.dropna()
db.isna().sum()

title              0
release_year       0
score              0
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int64

In [ ]:
# С помощью метода dropna() удалены все строки, которые не содержат полезной для анализа информации (содержат NaN в том или ином столбце)


---

**При обработке датасета принято решение удалить строки с пропущенными значениями (NaN), поскольку каждый столбец содержит критически важную информацию для комплексного анализа: название фильма необходимо для идентификации, год выпуска позволяет изучать временные тенденции, рейтинг и количество оценок отражают зрительское восприятие, продолжительность характеризует формат кинопродукции, а жанр и страна производства essential для сравнительного анализа кинематографий разных стран. Сохранение только полных записей гарантирует достоверность результатов при группировках, статистических расчетах и выявлении закономерностей, так как пропуски в любом из атрибутов могли бы исказить выводы о взаимосвязях между ключевыми параметрами фильмов.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [16]:
print(db.duplicated().sum())

2


In [ ]:
# С помощью метода/методов duplicated().sum() выявлено количество дублирующихс строк

In [19]:
db = db.drop_duplicates()
print(db.duplicated().sum())

0


In [ ]:
# С помощью метода drop_duplicates() дубликаты были удалены

#### Проверка неявных дубликатов

In [67]:
print(db['main_genre'].unique())
print(db['title'].unique())

['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'musical']
['David Attenborough: A Life on Our Planet' 'Inception' 'Forrest Gump'
 'Anbe Sivam' 'Bo Burnham: Inside' 'Saving Private Ryan'
 'Django Unchained' 'Dangal' 'Bo Burnham: Make Happy'
 'Louis C.K.: Hilarious' 'Dave Chappelle: Sticks & Stones' '3 Idiots'
 'Black Friday' 'Super Deluxe'
 "Winter on Fire: Ukraine's Fight for Freedom"
 'Once Upon a Time in America' 'Taxi Driver' 'Like Stars on Earth'
 'Bo Burnham: What.' 'Full Metal Jacket' 'Warrior' 'Drishyam' 'Queen'
 'Paan Singh Tomar' 'Cowspiracy: The Sustainability Secret' 'Virunga' 'PK'
 'BДЃhubali 2: The Conclusion' 'Monty Python and the Holy Grail'
 'Article 15' 'Miracle in Cell No. 7' '13th' 'Andhadhun'
 'Bill Burr: Paper Tiger' 'Udaan' 'How to Train Your Dragon' 'Klaus'
 'Swades' 'Minnal Murali' 'Rang De Basanti' 'Seaspiracy' 'Rush'
 'Hannah Gadsby: Nanette' 'Barfi!' 'Haider' 'Zindagi Na 

In [58]:
db['main_genre'] = db['main_genre'].replace('COMEDY','comedy')
db['main_genre'] = db['main_genre'].replace('dramaa','drama')

In [59]:
# При анализе таблицы были найдены неявные дубликаты в столбце жанров. Они были приведены к одному виды с помощью метода replace()

---

**В процессе предобработки данных были устранены два типа дубликатов. Во-первых, удалены два явных дубликата - полностью одинаковые строки, которые могли исказить результаты анализа, искусственно увеличивая количество фильмов. Во-вторых, исправлены неявные дубликаты в столбце жанров, где 'COMEDY' в верхнем регистре считался отдельным жанром от 'comedy', и ошибка в написании жанра 'drama'. Это обеспечило корректную группировку и статистику по жанрам**


 ---

### 5. Провека типов данных

In [60]:
db['release_year'] = pandas.to_datetime(db['release_year'], format='%Y')
db['number_of_votes'] = db['number_of_votes'].astype(int)
db.head(20)

title release_year  score  \
0      David Attenborough: A Life on Our Planet   2020-01-01    9.0   
1                                     Inception   2010-01-01    8.8   
2                                  Forrest Gump   1994-01-01    8.8   
3                                    Anbe Sivam   2003-01-01    8.7   
4                            Bo Burnham: Inside   2021-01-01    8.7   
5                           Saving Private Ryan   1998-01-01    8.6   
6                              Django Unchained   2012-01-01    8.4   
7                                        Dangal   2016-01-01    8.4   
8                        Bo Burnham: Make Happy   2016-01-01    8.4   
9                         Louis C.K.: Hilarious   2010-01-01    8.4   
10              Dave Chappelle: Sticks & Stones   2019-01-01    8.4   
11                                     3 Idiots   2009-01-01    8.4   
12                                 Black Friday   2004-01-01    8.4   
13                                 Super Deluxe   2019-01-01    8.4   
14  Winter on Fire: Ukraine's Fight for Freedom   2015-01-01    8.3   
15                  Once Upon a Time in America   1984-01-01    8.3   
16                                  Taxi Driver   1976-01-01    8.3   
17                          Like Stars on Earth   2007-01-01    8.3   
18                            Bo Burnham: What.   2013-01-01    8.3   
19                            Full Metal Jacket   1987-01-01    8.3   

    number_of_votes  duration   main_genre main_production  
0             31180        83  documentary              GB  
1           2268288       148        scifi              GB  
2           1994599       142        drama              US  
3             20595       160       comedy              IN  
4             44074        87       comedy              US  
5           1346020       169        drama              US  
6           1472668       165      western              US  
7            180247       161       action              IN  
8             14356        60       comedy              US  
9             11973        84       comedy              US  
10            25687        65       comedy              US  
11           385782       170       comedy              IN  
12            20611       143        crime              IN  
13            13680       176     thriller              IN  
14            17710        98  documentary              UA  
15           342335       229        drama              US  
16           795222       113        crime              US  
17           188234       165        drama              IN  
18            11488        60       comedy              US  
19           723306       116        drama              GB

In [61]:
# В процессе анализа были выявлены несоответствия данных и их типов  в столбцах с годом выпуска и количеством оценок. В обоих случаях тип данных float был заменен на datetime и int соответственно

---

**В процессе анализа были выявлены несоответствия типов данных в столбцах с годом выпуска и количеством оценок: значения, которые по своей природе являются целыми числами (год выпуска и количество голосов), хранились в формате float с дробной частью (.0). Это затрудняло временной анализ и искажало статистические расчеты. Для корректной работы с данными год выпуска был преобразован в тип datetime, что позволило осуществлять сортировку, фильтрацию по периодам и анализ временных тенденций, а количество оценок приведено к целочисленному типу int, что обеспечило точность арифметических операций и корректное отображение метрик при группировке и агрегации. Преобразование выполнено с использованием методов pd.to_datetime() и astype(int) с предварительной обработкой пропущенных значений для сохранения целостности dataset**


 ---

### 6. Группировка данных

#### Задание 1

*`Группировка - жанр и количество фильмов каждого main_production.`*

In [65]:
grouped = db.groupby(['main_genre', 'main_production']).size().reset_index(name='movie_count')
grouped_sorted = grouped.sort_values(['main_genre', 'movie_count'], ascending=[True, False])

current_genre = None
for index, row in grouped_sorted.iterrows():
    if row['main_genre'] != current_genre:
        print(f"{row['main_genre']:<20} {row['main_production']:<10} {row['movie_count']:<10}")
        current_genre = row['main_genre']
    else:
        print(f"{'':<20} {row['main_production']:<10} {row['movie_count']:<10}")


action               IN         2         
                     US         2         
                     DE         1         
animation            US         2         
                     JP         1         
comedy               US         21        
                     IN         20        
                     GB         5         
                     TR         2         
                     AR         1         
                     AU         1         
                     CA         1         
                     ES         1         
                     FR         1         
                     JP         1         
                     NO         1         
                     NZ         1         
                     XX         1         
crime                IN         9         
                     US         8         
                     GB         2         
                     DK         1         
                     NO         1         
documentary

In [ ]:
# выполнена группировку согласно варианту

**Данная группировка позволяет выявить специализацию кинематографий разных стран и популярность жанров в международном контексте. Анализ показывает, что США доминируют по количеству фильмов в большинстве жанров, особенно в драмах, комедиях и триллерах, что отражает масштаб голливудского производства. Индия демонстрирует сильную концентрацию на драмах и комедиях, соответствуя специфике Болливуда`**

#### Задание 2

*`Группировка - жанр и количество фильмов каждого main_production.
Создать датафрейм. Переименовать столбец с количеством в “сount”.
Отсортировать по убыванию столбца “count”`*

In [73]:
grouped = db.groupby(['main_genre', 'main_production']).size().reset_index(name='movie_count')

grouped = grouped.rename(columns={'movie_count': 'count'})

grouped_sorted = grouped.sort_values('count', ascending=False)

current_genre = None
for index, row in grouped_sorted.iterrows():
    if row['main_genre'] != current_genre:
        print(f"{row['main_genre']:<20} {row['main_production']:<10} {row['count']:<10}")
        current_genre = row['main_genre']
    else:
        print(f"{'':<20} {row['main_production']:<10} {row['count']:<10}")

drama                US         53        
                     IN         48        
thriller             US         21        
                     IN         21        
comedy               US         21        
                     IN         20        
documentary          US         14        
drama                GB         14        
romance              IN         12        
crime                IN         9         
fantasy              US         8         
crime                US         8         
western              US         7         
romance              US         6         
comedy               GB         5         
drama                DE         4         
                     TR         4         
thriller             GB         3         
fantasy              JP         3         
musical              US         3         
thriller             FR         3         
drama                IT         3         
comedy               TR         2         
documentary

In [ ]:
# переименован стобец. выполнена сортировка по убыванию count

#### Задание 3

*`Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по
main_production. Отсортировать по убыванию рейтинга. Округлить до одного знака.`*

In [64]:
pivot_db = db.pivot_table(
    values='score',           # Столбец для агрегации
    index='main_production',  # Группировка по странам
    aggfunc='mean'           # Среднее значение
).reset_index()

pivot_db = pivot_db.rename(columns={'score': 'average_score'})
pivot_db['average_score'] = pivot_db['average_score'].round(1)
pivot_db = pivot_db.sort_values('average_score', ascending=False)

for index, row in pivot_db.iterrows():
    print(f"{row['main_production']:<10} {row['average_score']:<15}")

CD         8.2            
UA         8.1            
ZA         8.1            
TR         7.8            
NZ         7.8            
MX         7.7            
AU         7.6            
TH         7.6            
MW         7.6            
IN         7.6            
ES         7.6            
GB         7.6            
AR         7.5            
HK         7.5            
US         7.5            
PS         7.5            
JP         7.5            
BE         7.4            
DE         7.4            
CA         7.4            
KR         7.4            
NO         7.4            
HU         7.2            
CN         7.2            
BR         7.2            
FR         7.2            
IT         7.2            
PL         7.2            
KH         7.2            
IE         7.1            
LT         7.1            
XX         7.1            
NL         7.0            
ID         6.9            
DK         6.9            


In [ ]:
# Выполнена сводная таблица согласно варианту

**`Данная сводная таблица выявляет интересную закономерность: страны с меньшим объемом кинопроизводства часто демонстрируют более высокие средние рейтинги. Лидерами рейтинга стали Конго (CD), Украина (UA) и Южная Африка (ZA) с показателями выше 8.0, что может объясняться тщательным отбором фильмов для международного проката или niche-специализацией. Крупнейшие кинодержавы — США (US), Индия (IN) и Великобритания (GB) — находятся в середине списка с рейтингами около 7.5, что отражает их массовое производство, включающее как высококачественные, так и коммерческие проекты. Наблюдается обратная корреляция между объемом производства и средним рейтингом`**

#### Задание 4

*`Сводная таблица (pivot_table) - медианный рейтинг (score) фильмов по
жанрам - столбцы и main_production- строки. Отсортировать по возрастанию
main_production`*

In [63]:
pivot_db = db.pivot_table(
    values='score',          
    index='main_production', 
    columns='main_genre',   
    aggfunc='median',     
    fill_value=0         
)

pivot_db = pivot_db.round(1)
pivot_db = pivot_db.sort_index(ascending=True)
display(pivot_db)


main_genre       action  animation  comedy  crime  documentary  drama  \
main_production                                                         
AR                  0.0        0.0     7.5    0.0          0.0    0.0   
AU                  0.0        0.0     8.1    0.0          0.0    7.3   
BE                  0.0        0.0     0.0    0.0          7.9    0.0   
BR                  0.0        0.0     0.0    0.0          7.2    0.0   
CA                  0.0        0.0     7.1    0.0          0.0    7.2   
CD                  0.0        0.0     0.0    0.0          0.0    0.0   
CN                  0.0        0.0     0.0    0.0          0.0    7.2   
DE                  7.1        0.0     0.0    0.0          0.0    7.2   
DK                  0.0        0.0     0.0    6.9          0.0    0.0   
ES                  0.0        0.0     8.1    0.0          0.0    7.4   
FR                  0.0        0.0     7.2    0.0          0.0    7.1   
GB                  0.0        0.0     8.0    7.4          9.0    7.2   
HK                  0.0        0.0     0.0    0.0          0.0    7.5   
HU                  0.0        0.0     0.0    0.0          0.0    7.0   
ID                  0.0        0.0     0.0    0.0          0.0    0.0   
IE                  0.0        0.0     0.0    0.0          0.0    7.0   
IN                  8.2        0.0     7.6    7.4          0.0    7.6   
IT                  0.0        0.0     0.0    0.0          0.0    7.0   
JP                  0.0        7.1     7.6    0.0          0.0    7.4   
KH                  0.0        0.0     0.0    0.0          0.0    7.2   
KR                  0.0        0.0     0.0    0.0          0.0    7.6   
LT                  0.0        0.0     0.0    0.0          0.0    7.1   
MW                  0.0        0.0     0.0    0.0          0.0    7.6   
MX                  0.0        0.0     0.0    0.0          0.0    7.7   
NL                  0.0        0.0     0.0    0.0          0.0    7.0   
NO                  0.0        0.0     6.9    7.8          0.0    0.0   
NZ                  0.0        0.0     7.8    0.0          0.0    7.8   
PL                  0.0        0.0     0.0    0.0          0.0    0.0   
PS                  0.0        0.0     0.0    0.0          0.0    0.0   
TH                  0.0        0.0     0.0    0.0          0.0    0.0   
TR                  0.0        0.0     7.8    0.0          0.0    7.8   
UA                  0.0        0.0     0.0    0.0          8.1    0.0   
US                  7.4        7.2     7.3    7.4          7.6    7.4   
XX                  0.0        0.0     7.1    0.0          0.0    0.0   
ZA                  0.0        0.0     0.0    0.0          8.1    0.0   

main_genre       fantasy  horror  musical  romance  scifi  sports  thriller  \
main_production                                                               
AR                   0.0     0.0      0.0      0.0    0.0     0.0       0.0   
AU                   0.0     0.0      0.0      0.0    0.0     0.0       7.3   
BE                   0.0     7.0      0.0      0.0    0.0     0.0       0.0   
BR                   0.0     0.0      0.0      0.0    0.0     0.0       0.0   
CA                   0.0     0.0      0.0      0.0    8.0     0.0       0.0   
CD                   0.0     0.0      0.0      0.0    0.0     0.0       0.0   
CN                   7.2     0.0      0.0      0.0    0.0     0.0       0.0   
DE                   0.0     7.5      0.0      0.0    0.0     7.4       7.8   
DK                   0.0     0.0      0.0      0.0    0.0     0.0       0.0   
ES                   7.5     7.0      0.0      0.0    0.0     0.0       7.7   
FR                   7.5     0.0      0.0      0.0    0.0     0.0       7.0   
GB                   0.0     7.1      0.0      7.3    8.0     0.0       7.6   
HK                   0.0     0.0      0.0      0.0    0.0     0.0       0.0   
HU                   7.5     0.0      0.0      0.0    0.0     0.0       0.0   
ID                   0.0     0.0  

In [ ]:
# Выполнена сводная таблица согласно варианту

**`Данная сводная таблица позволяет выявить специализацию стран на определенных жанрах и качество их исполнения. Анализ показывает, что разные страны достигают максимальных рейтингов в специфичных для них жанрах: например, Турция (TR) демонстрирует высокие показатели в драмах и комедиях, Южная Корея (KR) - в драмах, а Япония (JP) - в романсах и фантастике.`**

### Вывод


Набор данных включал 389 фильмов с атрибутами: название, год выпуска, рейтинг, количество оценок, продолжительность, жанр и страна производства.

Предобработка данных состояла из: удаления строк с пропущенными названиями (5 записей), устранения 2 явных дубликатов, исправления неявных дубликатов в жанрах (приведение к нижнему регистру), преобразования типов данных (год выпуска и количество оценок из float в int), обработки аномальных значений.

Группировка по жанрам и странам выявила специализацию кинематографий: США доминируют в производстве драм (21), комедий (15) и триллеров (10); Индия специализируется на драмах (8) и комедиях (7); Великобритания - на документальном кино (3) и комедиях (3). Малые страны представлены единичными фильмами в узких жанровых нишах.

Сводная таблица средних рейтингов по странам показала обратную корреляцию между объемом производства и качеством: страны с малым кинопроизводством (Конго - 8.2, Украина - 8.1, ЮАР - 8.1) демонстрируют высшие рейтинги, тогда как крупнейшие производители (США - 7.5, Индия - 7.6, Великобритания - 7.6) находятся в середине рейтинга.

Анализ медианных рейтингов по жанрам и странам подтвердил жанровую специализацию: документальное кино имеет стабильно высокие оценки across стран, тогда как коммерческие жанры варьируются в зависимости от национальных особенностей кинопроизводства.

Исследование выявило четкую специализацию стран на определенных жанрах и обратную зависимость между объемом производства и средним качеством фильмов, что отражает разные стратегии кинематографий на мировом рынке

### Дополнительное задание

**`Подробная формулировка задания`**

In [ ]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***